### PR TITLE
Charmrun: Refactor FT process restart to share code with normal launch

### PR DESCRIPTION
**Pull request by evan-charmworks**
This allows ++local and ++debug to apply to restarted processes.

Diff best viewed with --ignore-all-space --color-moved.

### CODE REVIEW

File: src/util/charmrun-src/charmrun.C [hunk 1]
```diff
@@ -4445,27 +4446,38 @@ static void envCat(char *dest, LPTSTR oldEnv)
 
 /* simple version of charmrun that avoids the sshd or charmd,   */
 /* it spawn the node program just on local machine using exec. */
-static void start_nodes_local(std::vector<nodetab_process> & process_table)
+struct local_nodestart
 {
   char cmdLine[10000];     /*Program command line, including executable name*/
```
Comment (matthiasdiener) [inline src/util/charmrun-src/charmrun.C]: Could change this into a std::string.

File: src/util/charmrun-src/charmrun.C [hunk 2]
```diff
@@ -5399,6 +5417,41 @@ static void start_nodes_mpiexec()
   /* all ssh_pid remain zero: skip finish_nodes */
 }
 
+static int finish_one_node(nodetab_process & p, int & retries, const char ** argv = arg_argv)
```
Comment (matthiasdiener) [inline src/util/charmrun-src/charmrun.C]: What's the meaning of the return value?

File: src/util/charmrun-src/charmrun.C [hunk 3]
```diff
@@ -5496,148 +5526,174 @@ static char *find_abs_path(const char *target)
...
-  /* copy environ and expanded to hold NETSTART and CmiNumNodes */
   int envc;
-  for (envc = 0; env[envc]; envc++)
-    ;
-  int extra = 0;
+  char **envp;
+  int n;
+
+  char **dparamp;
+  std::vector<char *> dparamv;
```
Comment (matthiasdiener) [inline src/util/charmrun-src/charmrun.C]: Could be a std::string as well.

File: src/util/charmrun-src/charmrun.C [hunk 4]
```diff
@@ -4445,27 +4446,38 @@ static void envCat(char *dest, LPTSTR oldEnv)
 
 /* simple version of charmrun that avoids the sshd or charmd,   */
 /* it spawn the node program just on local machine using exec. */
-static void start_nodes_local(std::vector<nodetab_process> & process_table)
+struct local_nodestart
```
Comment (matthiasdiener) [inline src/util/charmrun-src/charmrun.C]: maybe change to class

File: src/util/charmrun-src/charmrun.C (hunk 1 above)
Comment (evan-charmworks) [reply src/util/charmrun-src/charmrun.C]: Done

File: src/util/charmrun-src/charmrun.C (hunk 2 above)
Comment (evan-charmworks) [reply src/util/charmrun-src/charmrun.C]: It's an idiom where each non-success return site uses a different non-zero value. The return value is only used in a loop checking that it non-zero.

File: src/util/charmrun-src/charmrun.C (hunk 3 above)
Comment (evan-charmworks) [reply src/util/charmrun-src/charmrun.C]: Not the way it is used.

File: src/util/charmrun-src/charmrun.C (hunk 4 above)
Comment (evan-charmworks) [reply src/util/charmrun-src/charmrun.C]: If everything inside is public I prefer using struct.

**Comment by evan-charmworks** [comment]
> Is this meant for 6.10 still?

It could be.

File: src/util/charmrun-src/charmrun.C (hunk 4 above)
Comment (matthiasdiener) [reply src/util/charmrun-src/charmrun.C]: cmdLine doesn't even need to be public.

File: src/util/charmrun-src/charmrun.C (hunk 3 above)
Comment (matthiasdiener) [reply src/util/charmrun-src/charmrun.C]: Yeah, this would require more refactoring.

**Comment by evan-charmworks** [comment]
Since the minor fixes from this were merged in #2555, let's save the change that would require testing for 6.11.

**Comment by evan-charmworks** [comment]
Needs re-review

**Pull request merged**